teps/0133-configure-default-resolver.md

```diff
@@ -0,0 +1,163 @@
+- The default `Resolver` can either be a built-in type (`hub`, `git`, `cluster`, and `bundle`) or a [Custom Resolver].
+
+### Use Cases
+#### Custom Resolver
+As a company cluster operator, I may have my own [Custom Resolver] which is used by most of the cluster users when resolving remote resources. I may want to simplify the user experience with such default values at the authoring time.
+
+#### Common Remote Resource Storage
+As a company cluster operator, all of the company's `Tasks` and `Pipelines` are stored in the same GitHub repo. I may not want to enforce the company users to specify the `Resolver` type for every `TaskRun` and `PipelineRun` (see the [issue #5907]).
+
+## Proposal
+To support the configurable default `Resolver` type, we propose to introduce a new field `default-resolver-type` in the [config-defaults.yaml][config-defaults]. 
+
+With the `default-resolver-type` field configured, authors can create `TaskRun` and `PipelineRun` without setting the `resolver` field in the yaml input. 
+
+For example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: TaskRun
+metadata:
+ name: demo-task-run
+spec:
+ taskRef:
+   #resolver: git
+   params:
+   - name: url
+     value: https://github.com/tektoncd/catalog.git
+   - name: revision
+     value: main
+   - name: pathInRepo
+     value: task/golang-build/0.1/golang-build.yaml
+```
+
+Here is another example where a `Pipeline` tries to resolve a remote `PipelineTask`:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: demo-pipeline
+spec:
+  tasks:
+  - name: post-message
+    taskRef:
+      #resolver: hub
+      params:
+      - name: kind
+        value: task
+      - name: name
+        value: github-add-comment
+      - name: version
+        value: "0.7"
+    params:
+    ...
+```
+
+If the `default-resolver-type` field is **not** configured, the Tekton Pipeline project should behave the same way as before (i.e. the *validation webhook* should fail with a missing `resolver` field).
+
+To keep this new feature as "opt-in", we propose **not** to specify a default value to this field. 
+
+``` yaml
+  # Setting this flag to a resolver type as the default resolver
+  default-resolver-type: ""
+```
+
+The resource author can overwrite the default `Resolver` by explicitly setting it in the yaml input.
+
+## Design Details
+### Populate Default Resolver
+We propose to populate the default value in the current `SetDefault` function of `TaskRunSpec` and `PipelineSpec`.
+
+Taking `TaskRunSpec` as an example:
+
+```go
+// SetDefaults implements apis.Defaultable
+func (trs *TaskRunSpec) SetDefaults(ctx context.Context) {
+  cfg := config.FromContextOrDefaults(ctx)
+  if trs.TaskRef != nil {
+    if trs.TaskRef.Kind == "" {
+      trs.TaskRef.Kind = NamespacedTaskKind
+    }
+    if trs.TaskRef.Name == "" && trs.TaskRef.Resolver == "" {
+      trs.TaskRef.Resolver = ResolverName(cfg.Defaults.DefaultResolverType)
+    }
+  }
+```
```

Comment:
maybe we need one more "}" to fix the formatting issue?

```suggestion
```go
// SetDefaults implements apis.Defaultable
func (trs *TaskRunSpec) SetDefaults(ctx context.Context) {
  cfg := config.FromContextOrDefaults(ctx)
  if trs.TaskRef != nil {
    if trs.TaskRef.Kind == "" {
      trs.TaskRef.Kind = NamespacedTaskKind
    }
    if trs.TaskRef.Name == "" && trs.TaskRef.Resolver == "" {
      trs.TaskRef.Resolver = ResolverName(cfg.Defaults.DefaultResolverType)
    }
  }
}
```
```

Reply:
Thanks for catching it! Fixed!